clean/post_officer_history.py

```diff
@@ -0,0 +1,255 @@
+                .str.replace(r"^Hammond$", "Hammond PD", regex=True)
+                .str.replace(r"^New Orleans$", "New Orleans PD")
+                .str.replace(r"^Lafayette City$", "Lafayette City PD", regex=True)
+                .str.replace(r"^Grand Isle$", "Grand Isle PD", regex=True)
+                .str.replace(r"^New Orleans Levee$", "New Orleans Levee PD", regex=True)
+                .str.replace(r"^Harbor PD$", "New Orleans Harbor PD", regex=True)
+                .str.replace(r"^Lafayette$", "Lafayette PD", regex=True)
+                .str.replace(r"^Louisiana State$", "Louisiana State PD", regex=True)
+                .str.replace(r"^Caddo Parish$", "Caddo SO", regex=True)
+            )
+    return df
+
+
+def drop_rows_with_only_one_agency(df):
+    return df[~((df.agency.notna()) & (df.agency_1.fillna("") == ""))]
+
+
+def drop_rows_missing_names(df):
+    return df[~((df.first_name == "") & (df.last_name == ""))]
+
+
+def generate_history_id(df):
+    stacked_agency_sr = df[
+        [
+            "agency",
+            "agency_1",
+            "agency_2",
+            "agency_3",
+            "agency_4",
+            "agency_5",
+            "agency_6",
+            "agency_7",
+            "agency_8",
+        ]
+    ].stack()
+
+    stacked_agency_df = stacked_agency_sr.reset_index().iloc[:, [0, 2]]
+    stacked_agency_df.columns = ["history_id", "agency"]
+
+    names_df = df[["first_name", "last_name", "middle_name"]].reset_index()
+    names_df = names_df.rename(columns={"index": "history_id"})
+
+    stacked_agency_df = stacked_agency_df.merge(names_df, on="history_id", how="right")
+
+    return stacked_agency_df
+
+
+def check_for_duplicate_uids(df):
+    uids = df.groupby(["uid"])["history_id"].agg(list).reset_index()
+
+    for l in uids["history_id"]:
+        result = all(element == l[0] for element in l)
+        if result:
+            continue
+        else:
+            raise ValueError("uid found in multiple history ids")
+
+    return df
+
+
+def clean():
+    df = (
+        pd.read_csv(deba.data("raw/post/post_officer_history.csv"))
+        .pipe(clean_officer_names)
+        .rename(columns={"officer_sex": "sex"})
+        .pipe(clean_sexes, ["sex"])
+        .pipe(split_names)
+        .pipe(clean_names, ["first_name", "middle_name", "last_name"])
+        .pipe(clean_post_agency_row)
+        .pipe(extract_agency)
+        .pipe(
+            names_to_title_case,
+            [
+                "agency",
+                "agency_1",
+                "agency_2",
+                "agency_3",
+                "agency_4",
+                "agency_5",
+                "agency_6",
+                "agency_7",
+                "agency_8",
+            ],
+        )
+        .pipe(clean_post_agency_column)
+        .pipe(drop_rows_with_only_one_agency)
+        .pipe(drop_rows_missing_names)
+        .pipe(generate_history_id)
+        .pipe(gen_uid, ["first_name", "middle_name", "last_name", "agency"])
```

Comment:
Are there no employee id that come with officer history data?

Reply:
Unfortunately not